### ISSUE TITLE
Legacy BootCamp Windows does not appear

### ISSUE THREAD

**Issue by supaklaw** [dakanji/RefindPlus#148]
### RefindPlus Version

v0.13.3.AC Release

### Device Type

Apple Mac

### Problem Description

After successfully installing refindplus with MyBootMgr which shows my 2 mac drives, and a linux install disk, it does not show a legacy Bootcamp drive (running Win 7).  Within Mac Os (El Capitan) he startup disk control panel does show the drive, although previously it had displayed a windows drive icon, now shows a windows folder.  I can boot to it directly from that control panel but would like the option to see and boot into the Bootcamp drive from refindplus.  I am still using soft bless at the moment, I didn't want to hard bless until the issue is resolved or more info is available.

MP 3,1.  nVidia GTX 780, running 10.11.6

i dont know how to run the debugger efi via mybootmgr, but will with some instructions

### Problem Point

Before starting a loader or tool

### Affected Items

Win OS Loader

### Debug Log
[23d07c3447.log](https://github.com/dakanji/RefindPlus/files/10689154/23d07c3447.log)
[DebugLog_ConfigFactory.log](https://github.com/dakanji/RefindPlus/files/10679516/DebugLog_ConfigFactory.log)

**Comment by supaklaw** [comment]
also after looking at other tickets, adding Disk 1 is SSD with the sole Win 7/Bootcamp installed on it.   Disk 2 is primary MacOs el capitan, Disk 3 is my backup carbon cloned MacOS el capitan, disk 4 currently empty and wanted to install Linux on it.  But windows is on it's own dedicated drive.

**Comment by startergo** [comment]
Try gptsync.efi. I am not sure it comes with RP, but it comes with rEFInd so you can grab it from there. Once you load it from the boot menu if there is a misalignment in the partitions it will offer to sync partitions automatically.

**Comment by supaklaw** [comment]
I drop that in MyBootMgr > EFI > tools folder before running ConfigFactory?

**Comment by startergo** [comment]
It does not matter when you put it. As long it is there prior to booting.

**Comment by dakanji** [comment]
Let's complete the diagnosis before you run anything.

> dont know how to run the debugger efi

RefindPlus has two efi files:
A DEBUG file for trouble shooting and a RELEASE file for day to day use. The DBG file always generates a log file when in use and the REL file does not.

MyBootMgr always sets things up with the DBG file active and unless you have done the step to switch to the REL file via the `SwtchRunMode` app, the debug logs will be sitting in your ESP. So use the `TweakESP` app to mount the ESP you deployed to (on your `Mac HD` disk), grab and latest log file there and share that.

If you have already done that step, use the `SwtchRunMode` app to change back to DBG and reboot to get the log. When everything is sorted, use it to go back to REL and delete the log files in the ESP etc as set out in the instructions.

**Comment by supaklaw** [comment]
[23d07c3447.log](https://github.com/dakanji/RefindPlus/files/10689154/23d07c3447.log)

**Comment by dakanji** [comment]
Thanks.

Please unzip the attached to your desktop, mount the ESP as before, delete the `BOOTx64` files under the `BOOT` folder, paste the downloaded files into the same folder, restart the Mac, and share the log.

 
[X269-BOOTx64.zip](https://github.com/dakanji/RefindPlus/files/10693001/X269-BOOTx64.zip)

**Comment by supaklaw** [comment]
[23d09s4518.log](https://github.com/dakanji/RefindPlus/files/10699536/23d09s4518.log)

**Comment by dakanji** [comment]
Please replace the previous file with the attached and shared the log.
[X270-BOOTx64.zip](https://github.com/dakanji/RefindPlus/files/10703329/X270-BOOTx64.zip)

Note that this runs significantly slower as it outputs a detailed debug log.

Thanks

**Comment by supaklaw** [comment]
[23d10s5419.log](https://github.com/dakanji/RefindPlus/files/10709635/23d10s5419.log)

**Comment by dakanji** [comment]
Thanks.

So it seems RefindPlus is unable to detect the MBR Table and/or the  associated Legacy Boot Code for the BootCamp volume.

It may be that GptSync can solve this problem indeed. To run this, you can get a copy in the `Other Binaries` folder if you download the RefindPlus package here on GitHub (https://github.com/dakanji/RefindPlus/releases). You will need to mount your ESP, drop a copy into the `EFI/BOOT/Tools` folder, add it to the `showtools` list in the config file and reboot.

instead of GptSync, I'll get rid of BootCamp and the [HybridMBR hack](https://www.rodsbooks.com/gdisk/hybrid.html) it uses all together and use a [Bare Metal](https://forums.macrumors.com/threads/bare-metal-windows-on-cmp.2250317) installation instead myself. The linked instructions are for Windows 10 but can be used for any other version as long as you have the ISO.

HybridMBR is a hack to allow running legacy and UEFI loaders on the same disk that Apple adopted to run Legacy Windows and Mac OS on the same disk. Given that you are not even running Windows on  a disk with Mac OS, you don't need HybridMBR and its problems in the first place.

For instance, it is the reason why your `eDrive` volumes, a Windows volume, seems to have Apple's HFS+ filesystem; which seems odd. You need to make the call but going bare metal, either using that VM based process or a physical Windows unit, would be my suggestion. This is sure to be problem free on RefindPlus or anything else.

**Comment by startergo** [comment]
>For instance, it is the reason why your eDrive volumes, a Windows volume, seems to have Apple's HFS+ filesystem; which seems odd.
Why does it seem odd? That is precisely how hybrid MBR works. Bootcamp creates an NTFS partition alongside an HFS+ partition converting one bit in the EFI partition so the GPT drive appears as an MBR drive to the Windows OS.

**Comment by supaklaw** [comment]
The eDrive volume is an Apple Techtool recovery volume I don't use, but is HFS+ filesystem I believe.  It was created in El Capitan, not in Windows.

**Comment by supaklaw** [comment]
I followed the instructions for inserting gptsync into the config file, and it is not working.

What is happening is the main screen is showing two "Restore Tags" buttons, (the gptsync_x64 line was inserted after restore tags in the config file) and upon selecting either it operates as the restore tag function, but the highlighted selection returns to the first. In other words both buttons function as restore tags, the gptsync option is not visible.

![IMG_0427](https://user-images.githubusercontent.com/124734923/218192730-94839797-8741-4ad6-b081-f54bf19b1271.jpg)

**Comment by dakanji** [comment]
> Why does it seem odd?
A Windows disk in HFS+ seems odd to me. Either way, all moot given the explanation given.

> The eDrive volume is an Apple Techtool recovery volume I don't use, but is HFS+ filesystem I believe. It was created in El Capitan, not in Windows.

I see. I thought it was a Windows eDrive (for Encryption).
I suppose the options have been laid out clearly for you to make a call on how to proceed. Recommendation, if you can, is to ditch the HybridMBR. You don't need it to start with since you only have Windows on that disk.

**Comment by dakanji** [comment]
> I followed the instructions for inserting gptsync into the config file, and it is not working.

You need to go over the process again. Here it is on my unit:
![Screen Shot 2023-02-10 at 23 49 04](https://user-images.githubusercontent.com/1128876/218195088-3d86f3c7-fcca-4b28-869c-095a2b03c25d.jpg)

**Comment by supaklaw** [comment]
[config.txt](https://github.com/dakanji/RefindPlus/files/10711107/config.txt)

here's the config file, take a look

**Comment by dakanji** [comment]
You need to use the tags listed in the section above the line. That is, `gptsync` and not `gptsync_x64`.

The latter is one of several possible names the binary can have but what is written as what to use in the config file is what must be used there

**Comment by startergo** [comment]
```
showtools shell, bootorder, gdisk, gptsync, memtest, mok_tool, about, hidden_tags, reboot, exit, firmware, fwupdate
```
or whatever your ```showtools``` line includes plus ```gptsync.efi```.
and drop ```gptsync.efi``` in the ```tools``` folder.

**Comment by dakanji** [comment]
> or whatever your showtools line includes plus gptsync.efi.

I already gave him what is actually to be used.
It is simply `gptsync`. Not `gptsync_x64` or `gptsync.efi`.
What is required is written there in the config file. Please follow that.

**Comment by startergo** [comment]
I saw the error and corrected it ```gptsync``` in the ```show tools```. Thanks.

**Comment by dakanji** [comment]
Noted, but it was just repeating stuff (momentarily wrongly) that had already been given.

**Comment by dakanji** [comment]
> You need to use the tags listed in the section above the showtools line. 

Leaving you to it and will check in later. Hope it works out.
If it doesn't or even if it does. I still think ditching the HybridMBR is best but you may prefer to keep it of course.

**Comment by supaklaw** [comment]
I updated the config and the utility shows up.  

Does this utility actually mount a drive?  Because it looked to me like it was trying to reformat my Mac drive to install windows on it?  I don't want to build a HybridMBR, or re-install Windows.  I'd like to simply mount the Win 7 SSD.

**Comment by startergo** [comment]
it is not reformatting anything just syncing the partitions so the NTFS boot loader can be recognized again.

**Comment by startergo** [comment]
gptsync  reads the GPT partition table on device and synchronises the legacy MBR partition
       table with the GPT partition table. The MBR partition table can hold only 4 partitions.

       Legacy bootloaders or operating systems that do not support GPT require an  MBR  partition
       table; this is typically the case when using GRUB or LILO on an Intel-based Apple machine.

**Comment by supaklaw** [comment]
ok so just go ahead and hit yes to it's attempt to sync.  If it succeeds the drive will show up in main screen, if it fails it won't, correct?

**Comment by supaklaw** [comment]
Ok, gave it a try, and it didn't work.  Now as per suggestions, I was unaware of terminology in describing the issue.

This is a bare metal Window 7 install on it's own disk.  I installed directly from a Win7 disc onto the SSD drive, and then manually updated Bootcamp drivers within Windows.  I've been booting back and forth from each OS with no issues, and have used refind in the past and it saw all the drives no problem.

My understanding of the suggestions is that this drive just cannot be seen by ReFindPlus as a bare metal install?

**Comment by startergo** [comment]
>and have used refind in the past and it saw all the drives no problem.

Can you retest with ```rEFInd``` to see if it finds Windows 7?

**Comment by startergo** [comment]
All you need to do is to copy the ```rEFInd_x64.efi``` to the ```EFI``` folder and rename it to ```bootx64.efi```.

**Comment by startergo** [comment]
You need to copy/paste ```config.conf`` and rename it to ```refind.conf```.

**Comment by supaklaw** [comment]
ok.  Will this still boot into my non-flashed PC graphics card?

**Comment by supaklaw** [comment]
renaming old bootx64.efix as before, saving a copy of config and renaming that refind.conf

**Comment by startergo** [comment]
Don’t you have an original Apple GPU?

**Comment by supaklaw** [comment]
I do if needed, the entire idea was to not use it.  I prefer keeping my nVidia gtx780 usable throughout boot process, hence the reason I don't want to use regular refind.

Those steps didn't result in a refind launch, it just booted normally to my Mac OS login.

**Comment by startergo** [comment]
> Those steps didn't result in a refind launch, it just booted normally to my Mac OS login.

Did you replace the original ```bootx64.efi``` from RefindPlus?

**Comment by supaklaw** [comment]
yes

**Comment by supaklaw** [comment]
renamed config file, and replaced the bootx64 file, bootblessed it and relaunched

**Comment by startergo** [comment]
>renamed config file, and replaced the bootx64 file, bootblessed it and relaunched

Try with ```OPTION``` key pressed at boot.

**Comment by startergo** [comment]
You can bless from the boot screen:
```CNTRL+ENTER``` will bless the selection.

**Comment by dakanji** [comment]
> This is a bare metal Window 7 install on it's own disk. I installed directly from a Win7 disc onto the SSD drive, and then manually updated Bootcamp drivers within Windows.

I see. Then it should ordinarily be expected work although I haven't actually come across this (I assume you mean you installed on the Mac).

In this case, you might want to try on rEFInd again since if that is thte case, it should do the same on RefindPlus. Note however, that you will need to use a different GPU in as the one you are currently using will not work with rEFInd.

> The eDrive volume is an Apple Techtool recovery volume I don't use, but is HFS+ filesystem I believe. It was created in El Capitan, not in Windows.

Going back to this. Interestingly, the tool (never used it myself) has created the partition, or the partition has along the line been changed to, a `Basic Data Partition (BDP)` or as it is fully known,  a `Microsoft Basic Data Partition`. This is just as odd on a Mac OS Partition as HFS+ is on a Windows one. I don't think it is directly related to anything going on now, and I can see it is indeed on another disk, but odd all the same. 

Anyway, do the rEFind test and report back.

**Comment by dakanji** [comment]
> the main screen is showing two "Restore Tags"

BTW, this has exposed a bug in the inherited `showtools` processing logic.
It recognises an invalid item in the list but does not quite handle this.
Will fix RefindPlus and push this upstream later.

**Comment by supaklaw** [comment]
The bootcamp drive was corrupted and went into infinite repair loop, but managed to pull it and get back into the Mac drive.  Running the diskutil list shows a 'suspicious MBR header', not sure if that was from the merge attempt or what happened.  
[diskutil.txt](https://github.com/dakanji/RefindPlus/files/10731045/diskutil.txt)

**Comment by dakanji** [comment]
Suspicious MBR is what the tool rightly calls HybridMBR which should perhaps be renamed as "DodgyMBR".
So nothing unusual in it showing that by it self but as previously advised, you do not want to be using DodgyMBR unless you absolutely must; which is not the case for you.

You should do your disk and yourself a favour and stick to using "TrustworthyMBR" ... unless under duress!

Anyway, as the issue appears to have been due to disk health issues, I will tag this accordingly.

**Comment by supaklaw** [comment]
Correct, your friends insistence on using the MBR hybrid tool managed to damage my disk but feel free to tag this as "user error and bad disk" rather than bad software, poor english language skills, and clueless tech support.   Buhbye.

**Comment by startergo** [comment]
There is no `HybridMBr merge` the tool syncs partitions which are out of sync. It cannot damage the partitions. You can ask the creator of rEFInd if you don't trust me.
